modules/navier_stokes/src/executioners/SIMPLE.C

```diff
@@ -0,0 +1,1106 @@
+  {
+    _console << "Pressure solution" << std::endl;
+    solution.print();
+    _console << "Pressure solution old" << std::endl;
+    solution_old.print();
+  }
+
+  // We will overwrite the old solution here
+  solution_old = solution;
+  system_in.setSolution(solution);
+  system_in.residualSetup();
+}
+
+std::vector<Real>
+SIMPLE::solveMomentumPredictor()
+{
+  // Temporary storage for the (flux-normalized) residuals form
+  // different momentum components
+  std::vector<Real> normalized_residuals;
+
+  // We can create this here with the assumption that every momentum component has the same number
+  // of dofs
+  auto zero_solution = _momentum_systems[0]->system().current_local_solution->zero_clone();
+
+  // If we use a segregated approach between momentum components as well, we need to solve
+  // the other equations. Luckily, the system matrix is exactly the same so we only need
+  // to compute right hand sides.
+  for (const auto system_i : index_range(_momentum_systems))
+  {
+    _problem.setCurrentNonlinearSystem(_momentum_system_numbers[system_i]);
+
+    // We will need the right hand side and the solution of the next component
+    NonlinearImplicitSystem & momentum_system =
+        libMesh::cast_ref<NonlinearImplicitSystem &>(_momentum_systems[system_i]->system());
+
+    PetscLinearSolver<Real> & momentum_solver =
+        libMesh::cast_ref<PetscLinearSolver<Real> &>(*momentum_system.get_linear_solver());
+
+    NumericVector<Number> & solution = *(momentum_system.solution);
+    NumericVector<Number> & rhs = *(momentum_system.rhs);
+    SparseMatrix<Number> & mmat = *(momentum_system.matrix);
+
+    auto diff_diagonal = solution.zero_clone();
+
+    // We plug zero in this to get the system matrix and the right hand side of the linear problem
+    _problem.computeResidualAndJacobian(*zero_solution, rhs, mmat);
+    // Sadly, this returns -b so we multiply with -1
+    rhs.scale(-1.0);
+
+    // Still need to relax the right hand side with the same vector
+    relaxMatrix(mmat, _momentum_equation_relaxation, *diff_diagonal);
+    relaxRightHandSide(rhs, solution, *diff_diagonal);
+
+    // The normalization factor depends on the right hand side so we need to recompute it for this
+    // component
+    Real norm_factor = computeNormalizationFactor(solution, mmat, rhs);
+
+    // Very important, for deciding the convergence, we need the unpreconditioned
+    // norms in the linear solve
+    KSPSetNormType(momentum_solver.ksp(), KSP_NORM_UNPRECONDITIONED);
+    // Solve this component. We don't update the ghosted solution yet, that will come at the end
+    // of the corrector step Setting the lineat tolerances and maximum iteration counts
+    _momentum_ls_control.real_valued_data["abs_tol"] = _momentum_l_abs_tol * norm_factor;
+    momentum_solver.set_solver_configuration(_momentum_ls_control);
+
+    // We solve the equation
+    // TO DO: Add options to this function in Libmesh to accept absolute tolerance
```

Comment:
I think this has been done recently

Reply:
by @grmnptr himself!

Reply:
Oh yeah, I forgot to remove the todos :P